### PR TITLE
[SPARK-14349] [SQL] Issue Error Messages for Unsupported Operators/DML/DDL in SQL Context.

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

Currently, the weird error messages are issued if we use Hive Context-only operations in SQL Context. 

For example, 
- When calling `Drop Table` in SQL Context, we got the following message:

```
Expected exception org.apache.spark.sql.catalyst.parser.ParseException to be thrown, but java.lang.ClassCastException was thrown.
```
- When calling `Script Transform` in SQL Context, we got the message:

```
assertion failed: No plan for ScriptTransformation [key#9,value#10], cat, [tKey#155,tValue#156], null
+- LogicalRDD [key#9,value#10], MapPartitionsRDD[3] at beforeAll at BeforeAndAfterAll.scala:187
```

Updates:
Based on the investigation from @hvanhovell , the root cause is `visitChildren`, which is the default implementation. It always returns the result of the last defined context child. After merging the code changes from @hvanhovell , it works! Thank you @hvanhovell !
#### How was this patch tested?

A few test cases are added. 

Not sure if the same issue exist for the other operators/DDL/DML. @hvanhovell

### CODE REVIEW

**Comment by gatorsmile** [comment]
In SQL Context, the plan is wrong if we use CREATE VIEW AS SELECT. We need to 
```scala
  sql("CREATE VIEW testView AS SELECT * FROM jt").explain(true)
```

```
== Parsed Logical Plan ==
'Project [*]
+- 'UnresolvedRelation `jt`, None

== Analyzed Logical Plan ==
intType: int, stringType: string, dateType: date, timestampType: timestamp, doubleType: double, bigintType: bigint, tinyintType: tinyint, decimalType: decimal(10,0), fixedDecimalType: decimal(5,1), binaryType: binary, booleanType: boolean, smallIntType: smallint, floatType: float, mapType: map<string,string>, arrayType: array<string>, structType: struct<f1:string,f2:int>
Project [intType#155,stringType#156,dateType#157,timestampType#158,doubleType#159,bigintType#160L,tinyintType#161,decimalType#162,fixedDecimalType#163,binaryType#164,booleanType#165,smallIntType#166,floatType#167,mapType#168,arrayType#169,structType#170]
+- SubqueryAlias jt
   +- Relation[intType#155,stringType#156,dateType#157,timestampType#158,doubleType#159,bigintType#160L,tinyintType#161,decimalType#162,fixedDecimalType#163,binaryType#164,booleanType#165,smallIntType#166,floatType#167,mapType#168,arrayType#169,structType#170] SimpleDDLScan(1,10,test1)

== Optimized Logical Plan ==
Relation[intType#155,stringType#156,dateType#157,timestampType#158,doubleType#159,bigintType#160L,tinyintType#161,decimalType#162,fixedDecimalType#163,binaryType#164,booleanType#165,smallIntType#166,floatType#167,mapType#168,arrayType#169,structType#170] SimpleDDLScan(1,10,test1)

== Physical Plan ==
WholeStageCodegen
:  +- Scan SimpleDDLScan(1,10,test1)[intType#155,stringType#156,dateType#157,timestampType#158,doubleType#159,bigintType#160L,tinyintType#161,decimalType#162,fixedDecimalType#163,binaryType#164,booleanType#165,smallIntType#166,floatType#167,mapType#168,arrayType#169,structType#170]
```

The expected one should be like
```
== Parsed Logical Plan ==
'CreateViewAsSelect CatalogTable(`testView`,CatalogTableType(VIRTUAL_VIEW),CatalogStorageFormat(None,None,None,None,Map()),List(),List(),List(),0,1459654853237,1459654853237,Map(),Some(SELECT * FROM jt),Some(SELECT * FROM jt)), false, false, CREATE VIEW testView AS SELECT * FROM jt
+- 'Project [*]
   +- 'UnresolvedRelation `jt`, None

== Analyzed Logical Plan ==

CreateViewAsSelect CatalogTable(`default`.`testview`,CatalogTableType(VIRTUAL_VIEW),CatalogStorageFormat(None,None,None,None,Map()),List(),List(),List(),0,1459654853237,1459654853237,Map(),Some(SELECT * FROM jt),Some(SELECT * FROM jt)), Project [intType#0,stringType#1,dateType#2,timestampType#3,doubleType#4,bigintType#5L,tinyintType#6,decimalType#7,fixedDecimalType#8,binaryType#9,booleanType#10,smallIntType#11,floatType#12,mapType#13,arrayType#14,structType#15], false, false

== Optimized Logical Plan ==
CreateViewAsSelect CatalogTable(`default`.`testview`,CatalogTableType(VIRTUAL_VIEW),CatalogStorageFormat(None,None,None,None,Map()),List(),List(),List(),0,1459654853237,1459654853237,Map(),Some(SELECT * FROM jt),Some(SELECT * FROM jt)), Project [intType#0,stringType#1,dateType#2,timestampType#3,doubleType#4,bigintType#5L,tinyintType#6,decimalType#7,fixedDecimalType#8,binaryType#9,booleanType#10,smallIntType#11,floatType#12,mapType#13,arrayType#14,structType#15], false, false

== Physical Plan ==
ExecutedCommand CreateViewAsSelect CatalogTable(`default`.`testview`,CatalogTableType(VIRTUAL_VIEW),CatalogStorageFormat(None,None,None,None,Map()),List(),List(),List(),0,1459654853237,1459654853237,Map(),Some(SELECT * FROM jt),Some(SELECT * FROM jt)), Project [intType#0,stringType#1,dateType#2,timestampType#3,doubleType#4,bigintType#5L,tinyintType#6,decimalType#7,fixedDecimalType#8,binaryType#9,booleanType#10,smallIntType#11,floatType#12,mapType#13,arrayType#14,structType#15], false, false
```

We must disable it in the parser to make sure CREATE VIEW is allowed only in Hive Context.

**Comment by gatorsmile** [comment]
The same issue was found for create table:
```SQL
CREATE EXTERNAL TABLE parquet_tab2(c1 INT, c2 STRING)
TBLPROPERTIES('prop1Key '= "prop1Val", ' `prop2Key` '= "prop2Val")
```
Strange errors are got as shown below.
```
scala.collection.immutable.Map$Map2 cannot be cast to org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
java.lang.ClassCastException: scala.collection.immutable.Map$Map2 cannot be cast to org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
```

**Comment by gatorsmile** [comment]
cc @yhuai @hvanhovell Code is ready for review. Thanks!

**Comment by hvanhovell** [comment]
@gatorsmile good catch!

Let me try to explain why this happens.  All `visit*` functions in the ANTLR4 generated visitor use the  `AbstractParseTreeVisitor.visitChildren(...)` function for the default implementation. See: https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/tree/AbstractParseTreeVisitor.java#L61-L75

This has both advantages and disadvantages. The disadvantage is that this will select the result of the last defined context child. In the create table example you give, the last child is `TBLPROPERTIES('prop1Key '= "prop1Val", ' `prop2Key` '= "prop2Val")`, this makes the visitor return a `Map[String, String]` instead of a `LogicalPlan`. The advantage is that this provides a very nice default implementation for rules with exclusive ORs (the `statement` rule for instance).

The current approach of excluding every illegal option is error prone and is quite code intense. I think it is easier to improve the visitors default behavior by limiting the 'visitChildren' to one and only one child.

**Comment by gatorsmile** [comment]
@hvanhovell I see. I think you solution is much concise. Let me try to fix it. Thank you! : )

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 1]
```diff
@@ -133,6 +133,16 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
     }
   }
 
+  override def visitCreateFileFormat(
+      ctx: CreateFileFormatContext): CatalogStorageFormat = withOrigin(ctx) {
+    // Create the predicate.
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: Predicate?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 2]
```diff
@@ -133,6 +133,16 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
     }
   }
 
+  override def visitCreateFileFormat(
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: Could you add some doc to this.

Do you know which Hive test fails on this?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: sorry, ...because of the copy and paste...

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala (hunk 2 above)
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: Nvm I found the test which fail.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: NP

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: Yeah. Multiple test cases failed. For example,
``` test("SPARK-8811: compatibility with array of struct in Hive") ``` in parquetSuites.scala

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala [hunk 3]
```diff
@@ -134,6 +134,18 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
...
 
   /**
+   * Create a [[CatalogStorageFormat]]. This is part of the [[CreateTableAsSelect]] command.
+   */
+  override def visitCreateFileFormat(
+      ctx: CreateFileFormatContext): CatalogStorageFormat = withOrigin(ctx) {
+    if (ctx.storageHandler == null) {
+      typedVisit[CatalogStorageFormat](ctx.fileFormat)
+    } else {
+      typedVisit[CatalogStorageFormat](ctx.storageHandler)
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: MINOR/NIT: you could call straight into `visitStorageHandler` here.

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: Sure, can I change the return type of `visitStorageHandler` to `CatalogStorageFormat`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 4]
```diff
@@ -332,6 +332,9 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
       case r: RunnableCommand => ExecutedCommand(r) :: Nil
 
+      case _: logical.ScriptTransformation =>
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: This is a quite far down the line for catching such an problem.

We can also throw an exception in the `CatalystSqlParser`s 'withScriptIOSchema' function. We would have to move the 'transform query spec' tests in the PlanParserSuite to Hive (HiveQlSuite).

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Sure, will do. Thanks!

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: @hvanhovell In `CatalystSqlParsers`, how can we know the current context is Hive/SQL Context?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala (hunk 3 above)
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: Sure - someone (me) forgot to put in the wrong type so start with :)

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala]: : ) Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala (hunk 4 above)
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: @gatorsmile you can assume that only the `HiveSqlParser` supports this, and that a `HiveSqlParser` (and its builder) are only used within a `HiveContext`.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: uh, I see. Thank you! Will do the change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 5]
```diff
@@ -411,11 +412,12 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
    * Create a (Hive based) [[ScriptInputOutputSchema]].
    */
   protected def withScriptIOSchema(
+      ctx: QuerySpecificationContext,
       inRowFormat: RowFormatContext,
       recordWriter: Token,
       outRowFormat: RowFormatContext,
       recordReader: Token,
-      schemaLess: Boolean): ScriptInputOutputSchema = null
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Undo this change and throw the exception here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 6]
```diff
@@ -183,6 +185,20 @@ class SparkSqlAstBuilder extends AstBuilder {
   }
 
   /**
+   * Unsupported operation in SQL Context.
+   */
+  override protected def withScriptIOSchema(
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Remove this override and throw the exception in the `AstBuilder`.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: uh, got it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Done. Thank you!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveQlSuite.scala [hunk 7]
```diff
@@ -201,6 +202,31 @@ class HiveQlSuite extends SparkFunSuite with BeforeAndAfterAll {
     assert(plan.children.head.asInstanceOf[Generate].generator.isInstanceOf[JsonTuple])
   }
 
+  test("transform query spec") {
+    val plan1 = parser.parsePlan("select transform(a, b) using 'func' as c, d from e")
```
Comment (hvanhovell) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveQlSuite.scala]: What is the reason for changing the test? Is HiveQlSuite written in a different style?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 8]
```diff
@@ -398,11 +412,15 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
...
+      schemaLess: Boolean): ScriptInputOutputSchema = {
+    throw new ParseException(
+      "Script Transform is not supported in SQLContext. Use a HiveContext instead", ctx)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Minor: I am being a bit pendantic here. CatalystSqlParser knows nothing of a SQLContext/HiveContext... You could just say it is not supported, but this is fine.

**Comment by hvanhovell** [comment]
@gatorsmile we are getting there. Two minor comments.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveQlSuite.scala (hunk 7 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveQlSuite.scala]: : ) Let me import all the dsl from Catalyst. I have no clue why the test cases in sql/hive does not do it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 8 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Yeah, users might call different parsers. Let me call it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 9]
```diff
@@ -351,7 +364,8 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
           string(script),
           attributes,
           withFilter,
-          withScriptIOSchema(inRowFormat, recordWriter, outRowFormat, recordReader, schemaLess))
+          withScriptIOSchema(
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Nit: one line (i.e. revert this).

**Comment by hvanhovell** [comment]
LGTM - pending rebase/merge and jenkins.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 9 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Thanks! After we adding `ctx`, it has 102 characters. Thus, have to split it into two lines.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Ok, my bad :)

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: : )

**Comment by gatorsmile** [comment]
retest this please

**Comment by hvanhovell** [comment]
Merging to master. Thanks!
